File: packages/go/luna/resources/payments/payshap.go
// Package payments provides South African payment gateway integrations.
package payments

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"regexp"
	"time"

	lunahttp "github.com/eclipse-softworks/luna-sdk-go/luna/http"
)

// SABank represents a South African bank
type SABank string

const (
	BankABSA      SABank = "absa"
	BankCapitec   SABank = "capitec"
	BankFNB       SABank = "fnb"
	BankNedbank   SABank = "nedbank"
	BankStandard  SABank = "standard"
	BankInvestec  SABank = "investec"
	BankDiscovery SABank = "discovery"
	BankTymeBank  SABank = "tymebank"
	BankAfrican   SABank = "african"
)

// PayShap provides PayShap real-time payment integration.
type PayShap struct {
	client *lunahttp.Client
	config PayShapConfig
}

// NewPayShap creates a new PayShap instance.
func NewPayShap(client *lunahttp.Client, config PayShapConfig) *PayShap {
	return &PayShap{
		client: client,
		config: config,
	}
}

// CreatePayment creates a PayShap payment request.
func (p *PayShap) CreatePayment(ctx context.Context, req PayShapPaymentRequest) (*PayShapPayment, error) {
	paymentID := fmt.Sprintf("ps_%d", time.Now().UnixMilli())

	expiryMinutes := req.ExpiryMinutes
	if expiryMinutes == 0 {
		expiryMinutes = 30
	}
	expiresAt := time.Now().Add(time.Duration(expiryMinutes) * time.Minute)

	// Generate QR code data
	qrData := map[string]interface{}{
		"type":       "payshap",
		"merchantId": p.config.MerchantID,
		"amount":     req.Amount,
		"reference":  req.Reference,
	}
	qrJSON, _ := json.Marshal(qrData)
	qrCode := base64.StdEncoding.EncodeToString(qrJSON)

	return &PayShapPayment{
		ID:       paymentID,
		Provider: "payshap",
		ShapID:   fmt.Sprintf("shp_%s", paymentID[len(paymentID)-8:]),
		Amount: Amount{
			Value:    int(req.Amount * 100),
			Currency: "ZAR",
		},
		Status:    StatusPending,
		Reference: req.Reference,
		QRCode:    qrCode,
		ExpiresAt: expiresAt,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}, nil
}

// GetPayment gets payment status.
func (p *PayShap) GetPayment(ctx context.Context, paymentID string) (*PayShapPayment, error) {
	return &PayShapPayment{
		ID:       paymentID,
		Provider: "payshap",
		ShapID:   fmt.Sprintf("shp_%s", paymentID[len(paymentID)-8:]),
		Amount: Amount{
			Value:    0,
			Currency: "ZAR",
		},
		Status:    StatusPending,
		Reference: paymentID,
		ExpiresAt: time.Now().Add(30 * time.Minute),
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}, nil
}

// CancelPayment cancels a pending payment.
func (p *PayShap) CancelPayment(ctx context.Context, paymentID string) (*PayShapPayment, error) {
	payment, _ := p.GetPayment(ctx, paymentID)
	payment.Status = StatusCancelled
	payment.UpdatedAt = time.Now()
	return payment, nil
}

// LookupShapID looks up a ShapID (payment proxy).
func (p *PayShap) LookupShapID(ctx context.Context, shapID string) (map[string]interface{}, error) {
	pattern := regexp.MustCompile(`^[a-zA-Z0-9@._-]+$`)
	isValid := pattern.MatchString(shapID) && len(shapID) >= 5

	result := map[string]interface{}{
		"valid": isValid,
	}

	if isValid {
		result["bank_name"] = "Sample Bank"
		result["account_holder_name"] = "Account Holder"
	}

	return result, nil
}

// GenerateReceiveQR generates a QR code for receiving payments.
func (p *PayShap) GenerateReceiveQR(ctx context.Context, amount *float64, reference string) (map[string]string, error) {
	shapID := fmt.Sprintf("shp_%s_%d", p.config.MerchantID, time.Now().Unix())

	qrData := map[string]interface{}{
		"type":       "payshap_receive",
		"shapId":     shapID,
		"merchantId": p.config.MerchantID,
		"amount":     amount,
		"reference":  reference,
	}
	qrJSON, _ := json.Marshal(qrData)

	return map[string]string{
		"qr_code": base64.StdEncoding.EncodeToString(qrJSON),
		"shap_id": shapID,
	}, nil
}

// ValidateBankAccount validates a South African bank account number format.
func (p *PayShap) ValidateBankAccount(accountNumber string, bankID SABank) bool {
	accountLengths := map[SABank][]int{
		BankABSA:      {10, 11},
		BankCapitec:   {10},
		BankFNB:       {10, 11, 12},
		BankNedbank:   {10, 11},
		BankStandard:  {9, 10, 11},
		BankInvestec:  {10},
		BankDiscovery: {10},
		BankTymeBank:  {10},
		BankAfrican:   {11},
	}

	validLengths, ok := accountLengths[bankID]
	if !ok {
		return false
	}

	// Extract digits only
	digitsOnly := regexp.MustCompile(`\D`).ReplaceAllString(accountNumber, "")

	for _, length := range validLengths {
		if len(digitsOnly) == length {
			return true
		}
	}
	return false
}

File: packages/python/luna/resources/payments/payshap.py
"""Luna SDK - PayShap Integration for Python."""

from __future__ import annotations

import base64
import json
from datetime import datetime, timedelta
from typing import Literal

from luna.http import HttpClient
from .types import (
    PayShapConfig,
    PayShapPaymentRequest,
    PayShapPayment,
    Amount,
)


# South African banks with PayShap support
SA_BANKS = {
    "ABSA": "absa",
    "CAPITEC": "capitec",
    "FNB": "fnb",
    "NEDBANK": "nedbank",
    "STANDARD_BANK": "standard",
    "INVESTEC": "investec",
    "DISCOVERY_BANK": "discovery",
    "TYMEBANK": "tymebank",
    "AFRICAN_BANK": "african",
}

SABank = Literal["absa", "capitec", "fnb", "nedbank", "standard", "investec", "discovery", "tymebank", "african"]


class PayShap:
    """PayShap real-time payment integration."""

    def __init__(self, client: HttpClient, config: PayShapConfig) -> None:
        self._client = client
        self._config = config

    async def create_payment(self, request: PayShapPaymentRequest) -> PayShapPayment:
        """Create a PayShap payment request."""
        payment_id = f"ps_{int(datetime.now().timestamp() * 1000)}"
        expiry_minutes = request.expiry_minutes or 30
        expires_at = datetime.now() + timedelta(minutes=expiry_minutes)

        # Generate QR code data
        qr_data = json.dumps({
            "type": "payshap",
            "merchantId": self._config.merchant_id,
            "amount": request.amount,
            "reference": request.reference,
        })
        qr_code = base64.b64encode(qr_data.encode()).decode()

        return PayShapPayment(
            id=payment_id,
            provider="payshap",
            shap_id=f"shp_{payment_id[-8:]}",
            amount=Amount(value=int(request.amount * 100), currency="ZAR"),
            status="pending",
            reference=request.reference,
            qr_code=qr_code,
            expires_at=expires_at.isoformat(),
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
        )

    async def get_payment(self, payment_id: str) -> PayShapPayment:
        """Get payment status."""
        return PayShapPayment(
            id=payment_id,
            provider="payshap",
            shap_id=f"shp_{payment_id[-8:]}",
            amount=Amount(value=0, currency="ZAR"),
            status="pending",
            reference=payment_id,
            expires_at=(datetime.now() + timedelta(minutes=30)).isoformat(),
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
        )

    async def cancel_payment(self, payment_id: str) -> PayShapPayment:
        """Cancel a pending payment."""
        payment = await self.get_payment(payment_id)
        payment.status = "cancelled"
        payment.updated_at = datetime.now().isoformat()
        return payment

    async def lookup_shap_id(self, shap_id: str) -> dict:
        """Lookup a ShapID (payment proxy)."""
        import re
        is_valid = bool(re.match(r"^[a-zA-Z0-9@._-]+$", shap_id)) and len(shap_id) >= 5

        return {
            "valid": is_valid,
            "bank_name": "Sample Bank" if is_valid else None,
            "account_holder_name": "Account Holder" if is_valid else None,
        }

    async def generate_receive_qr(
        self,
        amount: float | None = None,
        reference: str | None = None,
    ) -> dict:
        """Generate a QR code for receiving payments."""
        shap_id = f"shp_{self._config.merchant_id}_{int(datetime.now().timestamp())}"

        qr_data = json.dumps({
            "type": "payshap_receive",
            "shapId": shap_id,
            "merchantId": self._config.merchant_id,
            "amount": amount,
            "reference": reference,
        })

        return {
            "qr_code": base64.b64encode(qr_data.encode()).decode(),
            "shap_id": shap_id,
        }

    def validate_bank_account(self, account_number: str, bank_id: SABank) -> bool:
        """Validate a South African bank account number format."""
        account_lengths: dict[SABank, list[int]] = {
            "absa": [10, 11],
            "capitec": [10],
            "fnb": [10, 11, 12],
            "nedbank": [10, 11],
            "standard": [9, 10, 11],
            "investec": [10],
            "discovery": [10],
            "tymebank": [10],
            "african": [11],
        }

        valid_lengths = account_lengths.get(bank_id)
        if not valid_lengths:
            return False

        digits_only = "".join(c for c in account_number if c.isdigit())
        return len(digits_only) in valid_lengths
